src/librustdoc/html/length_limit.rs

```diff
@@ -0,0 +1,88 @@
+//! See [`HtmlWithLimit`].
+
+use std::ops::ControlFlow;
+use std::fmt::Write;
+
+use crate::html::escape::Escape;
+
+/// A buffer that allows generating HTML with a length limit.
+///
+/// This buffer ensures that:
+///
+/// * all tags are closed,
+/// * only the most recently opened tag is closed,
+/// * no tags are left empty (e.g., `<em></em>`) due to the length limit being reached,
+/// * all text is escaped.
+#[derive(Debug)]
+pub(super) struct HtmlWithLimit {
```

Comment:
Would you like me to add some unit tests for `HtmlWithLimit`?

Reply:
Yes!